pkg/runtime/serializer/recognizer/recognizer.go

```diff
@@ -27,18 +27,38 @@ import (
 
 type RecognizingDecoder interface {
 	runtime.Decoder
+	// AccurateRecognizer should be true if this decoder can unambigously determine
+	// whether the content is known to it. Allows RecognizingDecoders to adjust certainty.
+	AccurateRecognizer() bool
+	// RecognizesData should return true if the input provided in the provided reader
+	// belongs to this decoder, or an error if the data could not be read or is ambiguous.
 	RecognizesData(peek io.Reader) (bool, error)
 }
 
+// NewDecoder creates a decoder that will attempt multiple decoders in an order defined
+// by:
+//
+// 1. The decoder implements RecognizingDecoder and returns true for AccurateRecognizer
+// 2. The decoder implements RecognizingDecoder and returns false for AccurateRecognizer
+// 3. All other decoders
+//
+// The order passed to the constructor is preserved within those priorities. Only the
+// last error encountered will be returned.
 func NewDecoder(decoders ...runtime.Decoder) runtime.Decoder {
 	recognizing, blind := []RecognizingDecoder{}, []runtime.Decoder{}
+	var ambiguous []runtime.Decoder
 	for _, d := range decoders {
 		if r, ok := d.(RecognizingDecoder); ok {
-			recognizing = append(recognizing, r)
+			if r.AccurateRecognizer() {
+				recognizing = append(recognizing, r)
+			} else {
+				ambiguous = append(ambiguous, r)
+			}
 		} else {
 			blind = append(blind, d)
 		}
 	}
+	blind = append(ambiguous, blind...)
```

Comment:
No, but I need to clarify why as a godoc - ambiguous can't guarantee that
no means no, so if we put it above other recognizers we get false
negatives.  The ordering of the input decoders is important, and the
guarantee that certainty comes first

On Wed, May 4, 2016 at 6:16 PM, Daniel Smith <notifications@github.com>
wrote:

> In pkg/runtime/serializer/recognizer/recognizer.go
> <https://github.com/kubernetes/kubernetes/pull/24710#discussion_r62124301>
> :
>
> >  		} else {
> >  			blind = append(blind, d)
> >  		}
> >  	}
> > +	blind = append(ambiguous, blind...)
>
> Shouldn't you append ambiguous with the recognizers instead?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/24710/files/ad0de6a6613f36d10064dd9ace7f65e6391353a2..f16ec0bf46975303284e1b9a7be1de4886526ddc#r62124301>
>